oggm/workflow.py

```diff
@@ -34,20 +34,33 @@ def _init_pool():
     return mp.Pool(cfg.PARAMS['mp_processes'], initializer=_init_pool_globals, initargs=(cfg_contents,))
 
 
+def _merge_dicts(*dicts):
+    r = {}
+    for d in dicts:
+        r.update(d)
+    return r
+
+
 class _pickle_copier(object):
     """Pickleable alternative to functools.partial,
     Which is not pickleable in python2 and thus doesn't work
     with Multiprocessing."""
 
-    def __init__(self, func, **kwargs):
+    def __init__(self, func, gdir_has_kwargs, **kwargs):
         self.call_func = func
+        self.gdir_has_kwargs = gdir_has_kwargs
         self.out_kwargs = kwargs
 
     def __call__(self, gdir):
-        return self.call_func(gdir, **self.out_kwargs)
+        if self.gdir_has_kwargs:
+            gdir, gdir_kwargs = gdir
+            gdir_kwargs = _merge_dicts(self.out_kwargs, gdir_kwargs)
+            return self.call_func(gdir, **gdir_kwargs)
+        else:
+            return self.call_func(gdir, **self.out_kwargs)
 
 
-def execute_entity_task(task, gdirs, **kwargs):
+def execute_entity_task(task, gdirs, gdir_has_kwargs=False, **kwargs):
     """Execute a task on gdirs.
```

Comment:
There are a couple problems with that signature:
It breaks compatibility with existing code, which is currently passing kwargs directly to the function.
With that, it would be impossible to potentially have common and per-gdir arguments at the same time.

Having the gdir and its kwargs in separate objects makes handling it in the function, and i believe in calling code as well, more complicated. You need to maintain two lists, match them up. A simple for gdir in gdirs loop won't work anymore.

I think the additional parameter can be safely dropped though, checking if the gdir is a tupple or not should be enough of a check, and would also enable a gdirs list with mixed normal gdirs and gdirs with kwargs.